### DISCUSSION TITLE
How to change penis color

### DISCUSSION THREAD

**Discussion by xArkailzx** [ErinaSugino/Starbound-SBR-Mod#64] [General]
I'm using the Avali compatibility mod. I tried to unpack the mod, change the textures, and repack it but it didn't worked. I don't know much about starbound modding at all, so I'm pretty much looking for the easiest solution. I opted to post my question here rather than elsewhere since I think I'll get the most help here.

Also, what is the sprite size limit for the penis ? I'd want to see if I can give a 'bigger' tool to my character (Enlarge penis for Actor1).

**Comment by ErinaSugino** [comment]
The genitals are treated like any other body part too. That is, they use color replacement directives from the actual in-world entity. I actually don't know what directives apply and from where from the top of my head - never had to work with actor building in detail yet - but I'd heavily assume they use the `bodyDirectives`. That means that the actual image in the sprite sheet is treated as a placeholder canvas, with each pixel actually being color-coded to be replaced by a specific different color that is defined in the bodyDirectives, for players mostly influenced by what palettes you chose on character creation. For example, afaik the Neki race uses the accessory color for the genitals which is the same replacement values as their collar uses.

Of course, if the source image uses color values that aren't defined to be replaced the original still applies. In which case your edit should have most likely worked. For more detailed troubleshooting you'd have to explain in more detail what exactly you have changed where and how.

As for changing the size of the genitals - there are mods out there that utilize this already, so it's possible. You can of course just replace the sprite images with bigger ones. The groin sprites are just overlayed over the body itself, so they use the maximum canvas size of 43x43 per image. However do keep in mind that you need to update the corresponsing mask files too (which define at what point in the image the genital is shown and where it's erased to emulate is going "inside" of something, otherwise there would be alot of clipping), and that the bigger you make the genitals, the weirder it will look in the end.

**Reply by xArkailzx** [reply]
But how can I change de bodyDirectives ? I'd want to use the avali trim color for it, but which file do I need to edit ? I'd guess I need to make a patch file, but I don't know what to change. As I said, I don't know much about starbound modding.
Edit: Changing the genitals size seems one hell of a ride. My priority is the color for now.

**Reply by ErinaSugino** [reply]
You can't change the body directives - atleast not without making a new character, doing error-prone manual savefile editing or using StarExtensions. In order to understand, you will need to snoop around the Avali mod'S species file and figure out how the race works. Usually at the bottom you will find three parameters `bodyColors`, `undyColors` and `hairColors`. These are the three color palettes, but depending on the race not all three might be populated. You will also find certain setting flags like `altOptionAsUndyColor`, which defines character creator behavior on how to apply the color palettes to the directives when you click through the options.

If you can read code and want a detailed understanding of how color palettes are assigned, you can take a look at `/scripts/sexbound/plugins/pregnant/baby.lua` at function `_convertBaByConfigToSpawnableNPC` - this method emulates *exactly* what Starbound is doing internally. Though if you want the easy answer, body directives are built from the body color, additionally the hair color if `hairColorAsBodySubColor` is set to true and additionally the undy color aswell, if `altOptionAsUndyColor` is set to true.

Now all that is left to do, after figuring out exactly which color palettes are used here, is to look at them in the species file and figure out which color values are the ones you want. The structure of data in the species file is a list of `colorValue` -> `replacementColorValue` on a pixel by pixel basis. So if you have found the trim colors you are looking for - that is your color palette to use on your changed gential sprites.
